spring-cloud-commons/src/test/java/reactor/core/publisher/FluxFirstNonEmptyEmittingTests.java

```diff
@@ -0,0 +1,187 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package reactor.core.publisher;
+
+import java.time.Duration;
+import java.util.Arrays;
+
+import org.junit.Test;
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscription;
+import reactor.core.CoreSubscriber;
+import reactor.core.Scannable;
+import reactor.test.StepVerifier;
+
+import static java.util.Collections.singletonList;
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * @author Tim Ysewyn
+ */
+public class FluxFirstNonEmptyEmittingTests {
+
+	@Test(expected = NullPointerException.class)
+	public void arrayNull() {
+		CloudFlux.firstNonEmpty((Publisher<Integer>[]) null);
+	}
+
+	@Test(expected = NullPointerException.class)
+	public void iterableNull() {
+		CloudFlux.firstNonEmpty((Iterable<Publisher<Integer>>) null);
+	}
+
+	@Test
+	public void firstWinner() {
+		StepVerifier
+				.create(CloudFlux.firstNonEmpty(Flux.range(1, 10), Flux.range(11, 10)))
+				.expectNext(1, 2, 3, 4, 5, 6, 7, 8, 9, 10).verifyComplete();
+	}
+
+	@Test
+	public void firstWinnerSecondEmpty() {
+		StepVerifier.create(CloudFlux.firstNonEmpty(Flux.range(1, 10), Flux.empty()))
+				.expectNext(1, 2, 3, 4, 5, 6, 7, 8, 9, 10).verifyComplete();
+	}
+
+	@Test
+	public void firstWinnerBackpressured() {
+		StepVerifier
+				.create(CloudFlux.firstNonEmpty(Flux.range(1, 10), Flux.range(11, 10)))
+				.thenRequest(5).expectNext(1, 2, 3, 4, 5).thenCancel()
+				.verifyThenAssertThat().hasNotDiscardedElements().hasNotDroppedElements()
+				.hasNotDroppedErrors();
+	}
+
+	@Test
+	public void secondWinner() {
+		StepVerifier
+				.create(CloudFlux.firstNonEmpty(Flux.never(), Flux.range(11, 10).log()))
+				.expectNext(11, 12, 13, 14, 15, 16, 17, 18, 19, 20).verifyComplete();
+	}
+
+	@Test
+	public void secondWinnerFirstEmpty() {
+		StepVerifier
+				.create(CloudFlux.firstNonEmpty(Flux.empty(), Flux.range(11, 10).log()))
+				.expectNext(11, 12, 13, 14, 15, 16, 17, 18, 19, 20).verifyComplete();
+	}
+
+	@Test
+	public void bothEmpty() {
+		StepVerifier.create(CloudFlux.firstNonEmpty(Flux.empty(), Flux.empty()))
+				.expectComplete().verifyThenAssertThat().hasNotDiscardedElements()
+				.hasNotDroppedElements().hasNotDroppedErrors();
+	}
+
+	@Test
+	public void neverAndEmpty() {
+		StepVerifier.create(CloudFlux.firstNonEmpty(Flux.never(), Flux.empty()))
+				.thenAwait(Duration.ofSeconds(2)).thenCancel().verifyThenAssertThat()
+				.hasNotDiscardedElements().hasNotDroppedElements().hasNotDroppedErrors();
+	}
+
+	@Test
+	public void firstEmitsError() {
+		RuntimeException ex = new RuntimeException("forced failure");
+		StepVerifier
+				.create(CloudFlux.firstNonEmpty(Flux.<Integer>error(ex), Flux.empty()))
+				.expectError(RuntimeException.class).verifyThenAssertThat()
```

Comment:
ideally you would check the message at least? can use `expectErrorMessage("forced failure")` for that.

Reply:
(same in the few tests below)